Add env var page to docs (#981)
open-telemetry/opentelemetry-go-instrumentation: docs/configuration.md

Reviewer: Some of these environment variables are documented in https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter or in https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables.
Maybe it will be better to point to those pages when applicable and keep here only the configuration specific to this project? If one of the env vars documentation is changed having a pointer to the official OTel docs will reduce the chance the docs here will become not up to date.

Reviewer 2: 1. Those env vars are mostly stable
2. Some of the env vars are not supported
3. Having everything in one page makes it simpler for the users
4. The spec and actual implementation can get out of sync

For reference in opentelemetry.io some existing automatic instrumentations are having dedicated configuration pages: 
- https://opentelemetry.io/docs/zero-code/net/configuration/
- https://opentelemetry.io/docs/languages/java/configuration/

For .NET, we are maintaining https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md and syncing https://opentelemetry.io/docs/zero-code/net/configuration during the release process.

# Configuration

> ## Configuration methods

Author: @MrAlias Should we add content here akin the [.NET docs](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#configuration-methods)? If so, can you provide rough content for me to add, or point to relevant resources?

Reviewer 2: The environmental variables listed below are used to configure the auto-instrumentation binary: https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/cli/main.go.

For the binary one can also set the logging level via `-logLevel` and enable recording telemetry from the OpenTelemetry default global implementation the global via `global-impl` CLI flags.

Users can also chose build their own Go automatic instrumentation binaries using https://pkg.go.dev/go.opentelemetry.io/auto and add support for more/other configurations.

Author: Should `global-impl` have a dash at the beginning (`-global-impl`)?

Reviewer 2: Yes, it should.

# Configuration

## Configuration methods

> ## Global settings

Author: @MrAlias Let me know if there is a better way to structure the env vars into separate sections. I based these mostly on what I found in the .NET docs.

Reviewer 2: lgtm

| Environment variable        | Description                                                                | Default value |
|-----------------------------|----------------------------------------------------------------------------|---------------|

> | `OTEL_GO_AUTO_TARGET_EXE`   | Sets the binary for the Go application to be instrumented.                 |               |

Author: Everywhere the default value column is blank, I could not find a default value. @MrAlias Please supply these if you know them.

> | `OTEL_GO_AUTO_TARGET_EXE`   | Sets the binary for the Go application to be instrumented.                 |               |

Reviewer 2: `OTEL_GO_AUTO_TARGET_EXE` is a required environment variable (by default it is empty).

> | `OTEL_GO_AUTO_GLOBAL`       | Enables the OpenTelemetry global implementation.                          |               |

Reviewer 2: Here is a more accurate description: https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/d730a91aaa5485951d9f8b3883b64936d021e693/cli/main.go#L62

> | `OTEL_GO_AUTO_GLOBAL`       | Enables the OpenTelemetry global implementation.                          |               |

Reviewer 2: Default: `false`

| `OTEL_EXPORTER_OTLP_PROTOCOL`               | Defines the OTLP exporter's transport protocol. Supported values: `grpc`, `http/protobuf`.                                                                                                                                                                                                                                                                                   | `http/protobuf`             |
| `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`        | Defines the OTLP exporter's transport protocol for the traces signal. Supported values: `grpc`, `http/protobuf`. The value of this environment variable takes precedence over `OTEL_EXPORTER_OTLP_PROTOCOL`.                                                                                                                                                                                                                                                             | `http/protobuf`             |

> | `OTEL_EXPORTER_OTLP_ENDPOINT`               | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment.                                                                                                                                | `https://localhost:4317`    |

Reviewer 2: Actually, I think the SDK docs are misleading. The default is `http://localhost:4318`  for http/protobuf protocol and `http://localhost:4317` for gRPC protocol (the connection is insecure by default).

> | `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`        | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment. The value of this environment variable takes precedence over `OTEL_EXPORTER_OTLP_ENDPOINT`.                                  | `https://localhost:4317`    |

Reviewer 2: Actually, I think the SDK docs are misleading. The default is `http://localhost:4318/v1/traces`  for http/protobuf protocol and `http://localhost:4317` for gRPC protocol (the connection is insecure by default).

| `OTEL_EXPORTER_OTLP_TRACES_INSECURE`        | Disables client transport security for the exporter's gRPC connection. You can use this only when an endpoint is provided without the `http or https` scheme. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_INSECURE`.                                                                                                                                  | `false`                     |
| `OTEL_EXPORTER_OTLP_HEADERS`                | Key-value pairs used as gRPC metadata associated with gRPC requests. The value must be represented in a format matching the W3C Baggage HTTP Header Content Format, except that additional semi-colon delimited metadata is not supported. Example value: `"key1=value1,key2=value2"`.                                                                                       | `none`                      |

> | `OTEL_EXPORTER_OTLP_TRACES_HEADERS`         | Key-value pairs used as gRPC metadata associated with gRPC requests. The value must be represented in a format matching the W3C Baggage HTTP Header Content Format, except that additional semi-colon delimited metadata is not supported. Example value: `"key1=value1,key2=value2"`. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_HEADERS`.           | `none`                      |

Reviewer 2: Default should be blank (unset). `none` is an actual possible value.

> | `OTEL_EXPORTER_OTLP_TIMEOUT`                | Maximum time in milliseconds the OTLP exporter waits for each batch export.                                                                                                                                                                                                                                                                                                  | `"10000"`                   |

Reviewer 2: ```suggestion
| `OTEL_EXPORTER_OTLP_TIMEOUT`                | Maximum time in milliseconds the OTLP exporter waits for each batch export.                                                                                                                                                                                                                                                                                                  | `10000`                   |
```

> | `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT`         | Maximum time in milliseconds the OTLP exporter waits for each batch export. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_TIMEOUT`.                                                                                                                                                                                                                   | `"10000"`                   |

Reviewer 2: ```suggestion
| `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT`         | Maximum time in milliseconds the OTLP exporter waits for each batch export. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_TIMEOUT`.                                                                                                                                                                                                                   | `10000`                   |
```

| `OTEL_EXPORTER_OTLP_TRACES_CLIENT_CERTIFICATE` | The filepath to the client certificate or chain of trust for the client's private key to use for mTLS communication in the PEM format. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_CLIENT_CERTIFICATE`.                                                                                                                                             | `none`                      |
| `OTEL_EXPORTER_OTLP_CLIENT_KEY`             | The filepath to the client's private key to use for mTLS communication in PEM format.                                                                                                                                                                                                                                                                                       | `none`                      |

> | `OTEL_EXPORTER_OTLP_TRACES_CLIENT_KEY`      | The filepath to the client's private key to use for mTLS communication in PEM format. The value of this variable takes precedence over `OTEL_EXPORTER_OTLP_CLIENT_KEY`.                                                                                                                                                                                                     | `none`                      |

Reviewer 2: Default should be blank (unset). `none` is an actual possible value.

# Configuration

## Configuration methods

Use the environment variables in the tables below to configure the [automatic instrumentation binary](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/cli/main.go).

Alternatively, you can add support for additional or different configurations by building your own Go automatic instrumentation binary using the [OpenTelemetry Go Automatic Instrumentation package](https://pkg.go.dev/go.opentelemetry.io/auto).

> ## Global settings

Reviewer: Maybe worth adding a section about instrumentation specific configuration? Currently the only example we have that I can think of is to include the SQL query in database/sql. But we are planning to have more in the future. If it is out of scope for this PR it's fine.

Reviewer 2: Can you share what config option is missing so that it can be added (here or separate PR)?

Reviewer: `OTEL_GO_AUTO_INCLUDE_DB_STATEMENT`

Reviewer: I think it might be clearer to have a separate section for instrumentation configuration (as the above option is not configuring the exporter but the instrumentation itself - collecting an extra attribute). Currently, the above env var is the only config option we have under that category so we can change it in the future once we'll have more options.

Author: IMO, it would be fine to create the section now with just the one entry. It might look a little funny to have just one config option, but I'd rather that than bury the env var in the wrong place and potentially confuse people about what it's for. But I don't have a ton of context, so we can go with whatever you recommend.

Reviewer 2: You can use https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#instrumentation-options for inspiration.

Author: I moved the env var to a new Instrumentation options section. As we fill in the new section with more env vars, we can add similar structure and detail to what @pellared linked to in the dotnet docs.

| Environment variable        | Description                                                                | Default value |
|-----------------------------|----------------------------------------------------------------------------|---------------|

> | `OTEL_GO_AUTO_TARGET_EXE`   | Sets the binary for the Go application to be instrumented. Required.       | Unset         |

Reviewer 2: Is the `Configuration methods` section not describing it already?

> | `OTEL_GO_AUTO_TARGET_EXE`   | Sets the binary for the Go application to be instrumented. Required.       | Unset         |

Reviewer: Right, just the word `Required` maybe should replaced/removed

> | `OTEL_GO_AUTO_TARGET_EXE`   | Sets the binary for the Go application to be instrumented. Required.       | Unset         |

Reviewer 2: @RonFed 

> Use the environment variables in the tables below to configure the [automatic instrumentation binary](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/cli/main.go).

and 

> It is only required when running the instrumentation as a standalone binary. 

Therefore, I think it is better to mention that it is required. Or am I missing something?

> | `OTEL_GO_AUTO_GLOBAL`       | Records telemetry from the OpenTelemetry default global implementation. As an alternative to using the environment variable, you can use the `-logLevel` CLI flag.    | `false`       |

Reviewer 2: ```suggestion
| `OTEL_GO_AUTO_GLOBAL`       | Records telemetry from the OpenTelemetry default global implementation. As an alternative to using the environment variable, you can use the `-global-impl` CLI flag.    | `false`       |
```

> | `OTEL_LOG_LEVEL`            | Sets the log level. Supported values: `none`, `error`, `warn`, `info`, `debug`. As an alternative to using the environment variable, you can use the `global-impl` CLI flag. | `info`        |

Reviewer 2: ```suggestion
| `OTEL_LOG_LEVEL`            | Sets the log level. Supported values: `none`, `error`, `warn`, `info`, `debug`. As an alternative to using the environment variable, you can use the `-logLevel` CLI flag. | `info`        |
```

| `OTEL_EXPORTER_OTLP_PROTOCOL`               | Defines the OTLP exporter's transport protocol. Supported values: `grpc`, `http/protobuf`.                                                                                                                                                                                                                                                                                   | `http/protobuf`             |
| `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`        | Defines the OTLP exporter's transport protocol for the traces signal. Supported values: `grpc`, `http/protobuf`. The value of this environment variable takes precedence over `OTEL_EXPORTER_OTLP_PROTOCOL`.                                                                                                                                                                 | `http/protobuf`             |

> | `OTEL_EXPORTER_OTLP_ENDPOINT`               | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment.                                                                                                                                | http/protobuf: `https://localhost:4318`, gRPC: `http://localhost:4317`    |

Reviewer 2: ```suggestion
| `OTEL_EXPORTER_OTLP_ENDPOINT`               | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment.                                                                                                                                | http/protobuf: `https://localhost:4318`, grpc: `http://localhost:4317`    |
```

> | `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`        | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment. The value of this environment variable takes precedence over `OTEL_EXPORTER_OTLP_ENDPOINT`.                                  | http/protobuf: `https://localhost:4318/v1/traces`, gRPC: `http://localhost:4317`    |

Reviewer 2: ```suggestion
| `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`        | The endpoint where the exporter sends the telemetry data. The value must contain a host, and can additionally include a port, scheme, and path. You can use either `http` or `https`. The value must not contain a query string or fragment. The value of this environment variable takes precedence over `OTEL_EXPORTER_OTLP_ENDPOINT`.                                  | http/protobuf: `https://localhost:4318/v1/traces`, grpc: `http://localhost:4317`    |
```
